NERDd/misp_receiver.py

```diff
@@ -0,0 +1,412 @@
+                if misp_url == ev['misp_instance'] and sighting['event_id'] == ev['event_id']:
+                    # correct 'misp_event' found, rewrite sightings is enough
+                    just_update = True
+            if just_update:
+                # ip record found, just rewrite sightings
+                tq.put_update_request("ip", ip_addr, [('array_upsert', 'misp_events',
+                                                       {'misp_instance': misp_url,
+                                                        'event_id': sighting['event_id']},
+                                                       [('set', 'sightings',
+                                                         get_sightings_for_nerd(sighting_list))])])
+                return
+        # ip address not even in NERD or not found correct 'misp_event', create new 'misp_event'
+        kwargs = {'values': ip_addr}
+        # find correct attribute to pass it to event creation
+        attributes = misp_inst.search(controller='attributes', **kwargs)['response']['Attribute']
+        for attrib_dict in attributes:
+            if attrib_dict['event_id'] == sighting['event_id']:
+                attrib = attrib_dict
+                break
+        else:
+            return
+        upsert_new_event(event, attrib, sighting_list)
+    except ConnectionError:
+        logger.error("Cannot connect to MISP instance!")
+
+
+def attrib_add_or_edit(ip_addr, event_id, attrib_id):
+    """
+    In case of attribute add, find corresponding event and fill it into NERD, in case of edit, remove old record before
+    inserting
+    :param ip_addr: value of attribute (has to be IP address)
+    :param event_id: id of event the attribute belongs to
+    :param attrib_id: id of attribute
+    :return: None
+    """
+    # get event from MISP, to which the attribute corresponds
+    try:
+        event = misp_inst.get(int(event_id))['Event']
+    except ConnectionError:
+        logger.error("Cannot connect to MISP instance!")
+        return
+    attrib = get_attribute_from_event(event, attrib_id)
+    if attrib is None:
+        return
+
+    if check_src_and_dst_one(event['Attribute'], ip_addr):
+        role = "src and dst at the same time"
+    else:
+        role = get_role_of_ip(attrib['type'])
+    # create new updated event and insert it to NERD
+    upsert_new_event(event, attrib, attrib.get('Sighting'), role)
+
+
+def receive_events():
+    """
+    Connect to MISP's ZeroMQ and listen for MISP's changes and react on them
+    :return: None
+    """
+    context = zmq.Context()
+    socket = context.socket(zmq.SUB)
+
+    socket.connect(misp_zmq)
+    time.sleep(1)
+    socket.setsockopt(zmq.SUBSCRIBE, b'')
+
+    while running_flag:
+        # wait for new message
+        message = socket.recv()
+
+        message = message.decode("utf-8")
+        # save json part of the message
+        json_message = json.loads("{" + message.split("{", 1)[1])
```

Comment:
How does the message look like? I suppose there's some text/data before the JSON part. Are you sure it can't contain the '{' character?
It would be good to explain it in a comment, if it's not too complicated.

Reply:
I can tell something from the code below, but it would still be a good idea to show an example of a message in a comment (for future readers of the code who don't know the message format)